addons/account/tests/common.py

```diff
@@ -20,38 +20,21 @@ class AccountTestInvoicingCommon(ProductCommon):
     # to override by the helper methods setup_country and setup_chart_template to adapt to a localization
     chart_template = False
     country_code = False
+    skip_common_data = False
 
     @classmethod
     def safe_copy(cls, record):
         return record and record.copy()
 
-    @staticmethod
-    def setup_country(country_code):
-
-        def _decorator(function):
-            @wraps(function)
-            def wrapper(self):
-                self.country_code = country_code.upper()
-                function(self)
-            return wrapper
-
-        return _decorator
-
-    @staticmethod
-    def setup_chart_template(chart_template):
-        def _decorator(function):
-            @wraps(function)
-            def wrapper(self):
-                self.chart_template = chart_template
-                function(self)
-            return wrapper
-
-        return _decorator
-
+    @data_depends(
+        'collect_company_accounting_data', 'setup_armageddon_tax',
+        '_create_product', 'copy_account', 'skip_common_data', 'country_code',
+        'chart_template',
+    )
     @classmethod
-    def setUpClass(cls):
-        super().setUpClass()
-
+    def setUpCommonData(cls):
```

Comment:
To be honest, adding another layer in tests looks messy and add more complexity. What about just relying on python classes? Just add a way to mark it as "executed only once" and it will do the job I think.
An example: `AccountTestInvoicingCommon` adds the ability to instanciate a new company with a specific country (`@AccountTestInvoicingCommon.setup_country`). Most of the tests in odoo are run using the US country. My idea is to just create a class `AccountTestCommonUS(AccountTestInvoicingCommon)` and mark it as to be run only once. That way, you have everything you want without changing anything in the existing code.

Reply:
Some test class in different modules will use a common Base Test class that will setup some data. It is especially True with l10n modules, but not only.

The setup is a big part of it, and will be used executed multiple times because it is impossible to keep the data cross test class, and even more cross module right now. 

This pr makes this possible. But there is nothing magic, an additional layer of complexity is needed to make it possible.

Reply:
This is still doable by design with the right tools: 
ProductCommon
AccountTestInvoicingCommon(ProductCommon)
TestSaleCommon(AccountTestInvoicingCommon)
QRPrintTest(AccountTestInvoicingCommon) (Switzerland)

ProductCommon could be executed only one.
AccountTestInvoicingCommon has to be executed twice: one for US, one for CH
TestSaleCommon is executed after AccountTestInvoicingCommon (US)
QRPrintTest is executed after AccountTestInvoicingCommon (CH)

To me, the only thing we need here is a better handling of overrides than affect the setupClass of a class in the hierarchy. That way, you know AccountTestInvoicingCommon has to be executed twice.